### ISSUE TITLE
Viewing DHCP fails

### ISSUE THREAD

**Issue by netman2048** [convergeone/freeztp#20]
Trying to run the "show ztp dhcpd leases all" command and it fails with the following output:

```
Traceback (most recent call last):
  File "/usr/bin/ztp", line 2766, in <module>
    interpreter()
  File "/usr/bin/ztp", line 2519, in interpreter
    console(tracking.show_dhcp_leases("all"))
  File "/usr/bin/ztp", line 2217, in show_dhcp_leases
    leases = dleases.get(mode)
  File "/usr/bin/ztp", line 2024, in get
    result.update({index: self._format_lease(lease)})
  File "/usr/bin/ztp", line 2035, in _format_lease
    "uid": lease.data["uid"]
KeyError: 'uid'
```

I can confirm that the DHCPD service is working as I do get IP's from the scope and the "show ztp dhcpd leases raw" command has output showing the leases.

Incidentally, I get a similar error when trying to provision a CSR1000V (which I don't think will work due to getting the serial number before auto-install runs:

```
[root@dev01 ~]# ztp run
2018-04-25 10:14:41:   interpreter: Command to run received. Calling start_tftp

2018-04-25 10:14:41:   start_tftp: Starting Up TFTPy

2018-04-25 10:23:23:   tracking_class.report: New transfer of (network-confg) from (10.50.50.135) detected

2018-04-25 10:23:23:   interceptor: Called. Checking the cache

2018-04-25 10:23:23:   file_cache.get: File (network-confg) requested by (10.50.50.135) not in cache

2018-04-25 10:23:23:   interceptor: Cache returned None. Running cfact.lookup procedure

2018-04-25 10:23:23:   cfact.lookup: Called. Checking filename (network-confg) and IP (10.50.50.135)

2018-04-25 10:23:23:   cfact.lookup: Requested filename matches the initialfilename. Returning True

2018-04-25 10:23:23:   interceptor: cfact.lookup returned (True)

2018-04-25 10:23:23:   interceptor: Returning ztp_dyn_file instantiated object

2018-04-25 10:23:23:   ztp_dyn_file: Instantiated as (<__main__.ztp_dyn_file instance at 0x22eaea8>)

2018-04-25 10:23:23:   cfact.request: Called with filename (network-confg) and IP (10.50.50.135)

2018-04-25 10:23:23:   cfact.request: Filename (network-confg) matches the configured initialfilename

2018-04-25 10:23:23:   cfact.request: Generated a TempID with cfact._generate_name: (ZTP-237FB7FFBA)

2018-04-25 10:23:23:   snmp_query._query_worker: Attempting SNMP Query

2018-04-25 10:23:23:   snmp_query._query_worker: Attempting SNMP Query

Traceback (most recent call last):
  File "/usr/bin/ztp", line 2766, in <module>
    interpreter()
  File "/usr/bin/ztp", line 2392, in interpreter
    start_tftp()
  File "/usr/bin/ztp", line 2355, in start_tftp
    server.listen(listenip="", listenport=69)
  File "/usr/lib/python2.7/site-packages/tftpy/TftpServer.py", line 166, in listen
    self.sessions[key].start(buffer)
  File "/usr/bin/ztp", line 1776, in start
    self.port)
  File "/usr/lib/python2.7/site-packages/tftpy/TftpStates.py", line 417, in handle
    rport)
  File "/usr/lib/python2.7/site-packages/tftpy/TftpStates.py", line 305, in handle
    self.context.dyn_file_func(self.context.file_to_transfer, raddress=raddress, rport=rport)
  File "/usr/bin/ztp", line 96, in interceptor
    zfile = ztp_dyn_file(afile, raddress, rport)
  File "/usr/bin/ztp", line 166, in __init__
    self.data = cfact.request(afile, raddress)
  File "/usr/bin/ztp", line 308, in request
    "Timestamp": time.time()
  File "/usr/bin/ztp", line 2242, in provision
    dhcpinfo = self.prov_get_mac(data["IP Address"])
  File "/usr/bin/ztp", line 2277, in prov_get_mac
    leases = dleases.get("current")
  File "/usr/bin/ztp", line 2018, in get
    result.update({index: self._format_lease(leases[lease])})
  File "/usr/bin/ztp", line 2035, in _format_lease
    "uid": lease.data["uid"]
KeyError: 'uid'
```

**Comment by PackeTsar** [comment]
Mike - I just pushed a branch called "v1.0.1" with a fix for this. Can you follow the below instructions to upgrade to it? Then try again.

- Get into your FreeZTP directory `cd freeztp`
- Update the repo `git pull`
- Change to the new branch `git checkout v1.0.1`
- Upgrade FreeZTP `python ztp.py upgrade`

**Comment by PackeTsar** [comment]
Also - The CSR1000v may provision just fine with a default config. I'm not sure how you could handle deploying a unique config. You may be able to find a SNMP OID to pull its MAC address and then use that as the unique identifier for a keystore. The serial number makes sense as a real-id to use for identification, but in reality, you can use anything that can be polled with SNMP. Just configure that SNMP OID as an snmpoid object.

**Comment by netman2048** [comment]
Good and bad results...  Upgrading to v1.0.1 fixed the show command issue.  I did try to use the SNMP OID for the MAC address but things error throwing some unicode errors.  It continues to work and doesn't fully exit.  It does appear to grab the data because later it shows it in the logs just before it 100% crashes.  If it is something you want to fix, I can create a new issue and input the specific error messages.

**Comment by PackeTsar** [comment]
Just post your results here and I’ll fix it real quick

**Comment by netman2048** [comment]
Here you go...  Almost seems like the dashes in the MAC address jacked it up.  The MAC is supposed to be: 00-0C-29-57-82-EB

[root@dev01 ~]# ztp run
2018-04-25 12:49:22:   interpreter: Command to run received. Calling start_tftp

2018-04-25 12:49:22:   start_tftp: Starting Up TFTPy

2018-04-25 12:56:01:   tracking_class.report: New transfer of (network-confg) from (10.50.50.135) detected

2018-04-25 12:56:01:   interceptor: Called. Checking the cache

2018-04-25 12:56:01:   file_cache.get: File (network-confg) requested by (10.50.50.135) not in cache

2018-04-25 12:56:01:   interceptor: Cache returned None. Running cfact.lookup procedure

2018-04-25 12:56:01:   cfact.lookup: Called. Checking filename (network-confg) and IP (10.50.50.135)

2018-04-25 12:56:01:   cfact.lookup: Requested filename matches the initialfilename. Returning True

2018-04-25 12:56:01:   interceptor: cfact.lookup returned (True)

2018-04-25 12:56:01:   interceptor: Returning ztp_dyn_file instantiated object

2018-04-25 12:56:01:   ztp_dyn_file: Instantiated as (<__main__.ztp_dyn_file instance at 0x235d3f8>)

2018-04-25 12:56:01:   cfact.request: Called with filename (network-confg) and IP (10.50.50.135)

2018-04-25 12:56:01:   cfact.request: Filename (network-confg) matches the configured initialfilename

2018-04-25 12:56:01:   cfact.request: Generated a TempID with cfact._generate_name: (ZTP-237FC5F922)

2018-04-25 12:56:02:   snmp_query._query_worker: Attempting SNMP Query
2018-04-25 12:56:02:   snmp_query._query_worker: Attempting SNMP Query

2018-04-25 12:56:02:   snmp_query._query_worker: Attempting SNMP Query


2018-04-25 12:56:02:   cfact.request: Generated a SNMP Request with TempID (ZTP-237FC5F922) and IP (10.50.50.135)

2018-04-25 12:56:02:   cfact.request: Returning the below config to TFTPy:
#########################
hostname ZTP-237FC5F922
!
snmp-server community secretcommunity RO
!
end
#########################

2018-04-25 12:56:02:   ztp_dyn_file: File size is 72 bytes

2018-04-25 12:56:02:   file_cache.store: Storing {1524675362.078414: {'ipaddr': '10.50.50.135', 'file': <__main__.ztp_dyn_file instance at 0x235d3f8>, 'filename': 'network-confg'}}

2018-04-25 12:56:02:   ztp_dyn_file.read: Called with size (512)

2018-04-25 12:56:02:   ztp_dyn_file.read: Returning position 0 to 512

2018-04-25 12:56:02:   interceptor: Called. Checking the cache

2018-04-25 12:56:02:   file_cache.get: File (network-confg) requested by (10.50.50.135) found in cache. Checking timeout.

2018-04-25 12:56:02:   file_cache.get: Cached file is still within timeout period. Resetting to 0 and returning

2018-04-25 12:56:02:   ztp_dyn_file.read: Called with size (512)

2018-04-25 12:56:02:   ztp_dyn_file.read: Returning position 0 to 512

2018-04-25 12:56:02:   ztp_dyn_file.close: Called. File closing.

2018-04-25 12:56:03:   tracking_class.report: New transfer of (ZTP-237FC5F922-confg) from (10.50.50.135) detected

2018-04-25 12:56:03:   interceptor: Called. Checking the cache

2018-04-25 12:56:03:   file_cache.get: File (ZTP-237FC5F922-confg) requested by (10.50.50.135) not in cache

2018-04-25 12:56:03:   interceptor: Cache returned None. Running cfact.lookup procedure

2018-04-25 12:56:03:   cfact.lookup: Called. Checking filename (ZTP-237FC5F922-confg) and IP (10.50.50.135)

2018-04-25 12:56:03:   cfact.lookup: TempID is (ZTP-237FC5F922)

2018-04-25 12:56:03:   snmp_query._query_worker: SNMP Query Successful (OID: WS-C3850_SERIAL_NUMBER). Host (10.50.50.135) responded with ()

2018-04-25 12:56:03:   cfact.lookup: Current SNMP Requests: ['ZTP-237FC5F922']

2018-04-25 12:56:03:   cfact.lookup: Seeing the suffix in the filename and the TempID in the SNMPRequests

2018-04-25 12:56:03:   cfact.lookup: The SNMP request is showing as completed

2018-04-25 12:56:03:   snmp_query._query_worker: SNMP Query Successful (OID: WS-C2960_SERIAL_NUMBER). Host (10.50.50.135) responded with ()

/usr/bin/ztp:282: UnicodeWarning: Unicode equal comparison failed to convert both arguments to Unicode - interpreting them as being unequal
  if serial in list(config.running["keyvalstore"]):
/usr/bin/ztp:289: UnicodeWarning: Unicode equal comparison failed to convert both arguments to Unicode - interpreting them as being unequal
  if serial in arraykeys:
Exception in thread Thread-9:
Traceback (most recent call last):
  File "/usr/lib64/python2.7/threading.py", line 812, in __bootstrap_inner
    self.run()
  File "/usr/lib64/python2.7/threading.py", line 765, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/bin/ztp", line 552, in _query_worker
    log("snmp_query._query_worker: SNMP Query Successful (OID: %s). Host (%s) responded with (%s)" % (oidname, self.host, str(response)))
UnicodeDecodeError: 'ascii' codec can't decode byte 0x82 in position 4: ordinal not in range(128)

2018-04-25 12:56:03:   cfact._default_lookup: Checking if a default-keystore is configured and ready...

2018-04-25 12:56:03:   cfact._default_lookup: A default keystore ID (DEFAULT_VALUES) is configured

2018-04-25 12:56:03:   cfact.lookup: The target ID is NOT in the Keystore or in an IDArray, but a default is configured

2018-04-25 12:56:03:   interceptor: cfact.lookup returned (True)

2018-04-25 12:56:03:   interceptor: Returning ztp_dyn_file instantiated object

2018-04-25 12:56:03:   ztp_dyn_file: Instantiated as (<__main__.ztp_dyn_file instance at 0x7fb5d003abd8>)

2018-04-25 12:56:03:   cfact.request: Called with filename (ZTP-237FC5F922-confg) and IP (10.50.50.135)

2018-04-25 12:56:03:   cfact.request: Filename (ZTP-237FC5F922-confg) does NOT match the configured initialfilename

2018-04-25 12:56:03:   cfact.request: Stripped filename to TempID (ZTP-237FC5F922)

2018-04-25 12:56:03:   cfact.request: Delaying keystore loopup for (1000) milliseconds

2018-04-25 12:56:04:   cfact.request: Seeing the suffix in the filename and the TempID in the SNMP Requests

2018-04-25 12:56:04:   cfact.request: SNMP Request says it has completed

2018-04-25 12:56:04:   cfact.request: SNMP request returned target ID ({u'WS-C2960_SERIAL_NUMBER': '', u'CSR1KV_G1_MAC': '\x00\x0c)W\x82\xeb', u'WS-C3850_SERIAL_NUMBER': ''})

Traceback (most recent call last):
  File "/usr/bin/ztp", line 2770, in <module>
    interpreter()
  File "/usr/bin/ztp", line 2396, in interpreter
    start_tftp()
  File "/usr/bin/ztp", line 2359, in start_tftp
    server.listen(listenip="", listenport=69)
  File "/usr/lib/python2.7/site-packages/tftpy/TftpServer.py", line 166, in listen
    self.sessions[key].start(buffer)
  File "/usr/bin/ztp", line 1776, in start
    self.port)
  File "/usr/lib/python2.7/site-packages/tftpy/TftpStates.py", line 417, in handle
    rport)
  File "/usr/lib/python2.7/site-packages/tftpy/TftpStates.py", line 305, in handle
    self.context.dyn_file_func(self.context.file_to_transfer, raddress=raddress, rport=rport)
  File "/usr/bin/ztp", line 96, in interceptor
    zfile = ztp_dyn_file(afile, raddress, rport)
  File "/usr/bin/ztp", line 166, in __init__
    self.data = cfact.request(afile, raddress)
  File "/usr/bin/ztp", line 348, in request
    "Timestamp": time.time()
  File "/usr/bin/ztp", line 2233, in provision
    self.provdb(current)
  File "/usr/bin/ztp", line 2298, in __call__
    self._write()
  File "/usr/bin/ztp", line 2305, in _write
    rawout = json.dumps(fulldb, indent=4, sort_keys=True)
  File "/usr/lib64/python2.7/json/__init__.py", line 250, in dumps
    sort_keys=sort_keys, **kw).encode(obj)
  File "/usr/lib64/python2.7/json/encoder.py", line 209, in encode
    chunks = list(chunks)
  File "/usr/lib64/python2.7/json/encoder.py", line 434, in _iterencode
    for chunk in _iterencode_dict(o, _current_indent_level):
  File "/usr/lib64/python2.7/json/encoder.py", line 408, in _iterencode_dict
    for chunk in chunks:
  File "/usr/lib64/python2.7/json/encoder.py", line 408, in _iterencode_dict
    for chunk in chunks:
  File "/usr/lib64/python2.7/json/encoder.py", line 408, in _iterencode_dict
    for chunk in chunks:
  File "/usr/lib64/python2.7/json/encoder.py", line 390, in _iterencode_dict
    yield _encoder(value)
UnicodeDecodeError: 'utf8' codec can't decode byte 0x82 in position 4: invalid start byte
[root@dev01 ~]#

**Comment by PackeTsar** [comment]
What OID did you use?

**Comment by netman2048** [comment]
1.3.6.1.2.1.2.2.1.6.1

**Comment by PackeTsar** [comment]
OK give it a try now. Looks like SNMP returns hex data even though it is labeled as a string. Had to put in some logic to try encoding it a few different ways. Use the same process to upgrade.

Also, you can test SNMP using `ztp request snmp-test <ip-address>` to get faster results on if SNMP works properly and what values it returns.

**Comment by netman2048** [comment]
Looks like it worked although I had two issue.  First time around I think the SNMP requests seemed to take too long so I'm not sure it got the response in time before it decided to use the default settings.  I think you have a timer to address that but gotta go back and look.  The second issue was I was expecting the MAC address with dashes since that is what an SNMP get showed but when I ran your snmp-test command, I saw it was formatted different so I just had to change my idarray.

On a side note, have you considered implementing a switch/command for a one time run configuration?  For example, if I have 100 switches to deploy, I have to program 100 keystores, 100 idarrays, 100 associations then after everything gets programmed, I have to go back in and remove all that config.  It would be nice if I supplied a config file that only applied for the duration of the session.  If a one time run setup wasn't desired, maybe an option to remove the relevant config after a successfully provisioning the device.

The other idea I had was as an alternate option to using set commands to set keystore values, what about using a CSV/XLS doc?  Use the column headers as the variable names and each row representing a different device.  The idarray value could be the first column so the switch could be identified.  I was working on an ACI provisioning script that read in a bunch of stuff from spreadsheet tabs and attempted to do a complete provisioning of a fabric.  Reading in settings from a spreadsheet worked fairly well.

Just some thoughts to streamline things a little bit.

**Comment by PackeTsar** [comment]
I'm curious about the "one time run" function you are talking about. How would you supply unique values for each device configuration? I have considered allowing the input of a list of values for a keystore and allowing that keystore to pull each value only once. Is that what you mean?

**Comment by netman2048** [comment]
Here's my thought.  I work for a VAR and many times I have projects where I'm generating a set of configs and when its all done, I don't need all those values anymore.  The way the program works now, I have to clean up after myself which I don't want to do because it takes time and I'm lazy hehe.  It would be easier for me to put unique values in a CSV/Excel, have the program match on the serial or MAC to find the right row, and grab the values to inject into the J2 template.  I did something similar with Ansible where I made a playbook taking advantage of a module that can read CSV/XLS to import facts to inject into the J2 template...however at that point I have to manually copy/paste the config in.  I like that your program can upgrade the image and do ZTP.

For what it's worth, I got an email blast from Ivan Pepelnjak this morning about one of his students doing a similar ZTP program using Ansible.  I haven't had a chance to check it out yet, but it looks like it uses Excel to import facts.  Looks like Ivan also linked to your project in the email which is cool for you.

**Comment by PackeTsar** [comment]
Hmm that makes sense. What if I allow the creation of dynamic keystore objects which will pull keystore type data from external sources? It would map data in the background as a keystore would shown below.

### A table like this

| ID        | name           | ip  |
| ------------- |:-------------:| -----:|
| FCN1234567 | SWITCH1 | 192.168.1.1 |
| FCN1234568 | SWITCH2 | 192.168.1.2 |
| FCN1234569 | SWITCH3 | 192.168.1.3 |

### Would act in ZTP like invisible keystores formatted like:
```
ztp set keystore FCN1234567 name SWITCH1
ztp set keystore FCN1234567 ip 192.168.1.1
#
ztp set keystore FCN1234568 name SWITCH2
ztp set keystore FCN1234568 ip 192.168.1.2
#
ztp set keystore FCN1234569 name SWITCH3
ztp set keystore FCN1234569 ip 192.168.1.3
```


The ZTP config would likely look something like
```
ztp set external-keystore MYEXCELSHEET type csv
ztp set external-keystore MYEXCELSHEET file "/etc/ztp/dynsheet/somecsv.csv"
```


I could also extend this to build in other external-keystore types like a Google Sheet. 
```
ztp set external-keystore MYGSHEET type gsheet
ztp set external-keystore MYGSHEET api-key blahblahblah
ztp set external-keystore MYGSHEET sheetid 1234567890
ztp set external-keystore MYGSHEET sheetname myswitches
```

Does this fit what you had in mind? I'm already working on building external integrations into v1.1.0 so ZTP can send external notifications when a switch comes up (so far it will likely be able to dump data to spark, google sheets, possibly slack and SMTP) so adding external data sources shouldn't be too much trouble.

**Comment by netman2048** [comment]
Yes that is the idea....  I'm just curious, why have you not attempted to use Cisco's newer ZTP implementation where the device will download a python script from TFTP (or wherever) and then attempt to execute it locally?  I started monkeying around with that but I think I'm hitting a bug in their cli module that they want you to use...at least on the CSR platform.

**Comment by PackeTsar** [comment]
TBH I haven't really played with it much. I'm planning to play around with that and PnP to see if there are any good ways to leverage them in FreeZTP.

**Comment by PackeTsar** [comment]
I'm updating the v1.1.0 requested enhancements with the CSV and Python script idea and closing this issue since it seems to have been fixed in v1.0.1
